### PR TITLE
[Fix] Fix bitstring/cost mispairing in simulated annealing

### LINKED ISSUE #230: Error in the bitstrings distribution of Simulated Annealing
### Short description

- Simulated Annealing returns the right bitstrings and the right costs, but pairs them up wrongly. Each cost belongs to a different bitstring 
- I think the error is after  `torch.unique(..., dim=0),` the line `energies = energies[inverse_indices]` reorders the costs backward. I'm still not sure. 
- In the current version. Any code that picks the best solution from SA by cost picks the wrong one. Measured on a 6-variable QUBO (code below to reproduce): you get a solution worth -7 while the -11 optimum is in the same distribution.
- Makes HybridSATabuSolver seed tabu search from the wrong bitstring

### What is the expected result?

_No response_

### What is the actual result?

```
`SA on a 6-variable QUBO, seed=2, top_k=8.

  bitstring    cost SA reports    real x^T Q x
  ---------------------------------------------
  010001           -7.0             -7.0
  010101           -9.0             -9.0
  100101           -8.0             -8.0
  101010           -8.0             -8.0
  101110           -8.0             -8.0
  110001           -7.0             -7.0
  110100          -11.0             -7.0   <-- mispaired
  110101           -7.0            -11.0   <-- mispaired

2 of 8 costs sit on the wrong bitstring.
Both columns hold the same numbers, only mispaired: True
That is why min(costs) looks fine and the bug hides.

  bitstrings[argmin(costs)] = 110100
    you believe it is worth   -11.0
    it is really worth         -7.0
    best in this same distribution  -11.0  (110101)`
```

### Steps/Code to reproduce

```
import torch

from qubosolver.classical_solver.simulated_annealing import simulated_annealing

Q = torch.tensor(
    [
        [-2.0, 1.0, 0.0, -1.0, 0.0, 0.0],
        [1.0, -3.0, 2.0, 0.0, 0.0, -1.0],
        [0.0, 2.0, -1.0, 1.0, -2.0, 0.0],
        [-1.0, 0.0, 1.0, -2.0, 1.0, 0.0],
        [0.0, 0.0, -2.0, 1.0, -1.0, 2.0],
        [0.0, -1.0, 0.0, 0.0, 2.0, -2.0],
    ],
    dtype=torch.float64,
)

def demo_the_bug(seed: int = 2) -> bool:
    """Run SA, then compare every reported cost against x^T Q x."""
    bitstrings, reported, _counts = simulated_annealing(
        Q=Q, top_k=8, max_iter=5000, initial_temp=4.0, final_temp=0.05, seed=seed
    )

    x = bitstrings.to(torch.float64)
    # Row-wise x^T Q x over the whole batch of bitstrings at once.
    true = (x @ Q * x).sum(dim=1)

    print(f"SA on a 6-variable QUBO, seed={seed}, top_k=8.\n")
    print("  bitstring    cost SA reports    real x^T Q x")
    print("  " + "-" * 45)
    for bits, r, t in zip(bitstrings.tolist(), reported.tolist(), true.tolist()):
        flag = "   <-- mispaired" if abs(r - t) > 1e-9 else ""
        print(f"  {''.join(map(str, bits))}         {r:>6.1f}           {t:>6.1f}{flag}")

  
    wrong = int((reported - true).abs().gt(1e-9).sum())
    print(f"\n{wrong} of {len(true)} costs sit on the wrong bitstring.")

    shuffled = torch.equal(reported.sort().values, true.sort().values)
    print(f"Both columns hold the same numbers, only mispaired: {shuffled}")
    print("That is why min(costs) looks fine and the bug hides.\n")

    return
```

### Tracebacks (optional)

_No response_

### Environment details (optional)

I'm using the negative-coefficients branch, so I discovered this error first in there. To be sure it was not something from that branch, I tested it with the latest PyPI release (0.8.0) and with the main  branch.

### Would you like to work on this issue?

None

### PR DESCRIPTION
**Pull request by abussy-pasqal**
Fixes #230

torch.unique(..., return_inverse=True) followed by energies[inverse_indices] performed a gather where a scatter was required: inverse_indices[i] names the unique-row slot original row i maps to, so aligning energies with unique_bits means writing energies into that slot, not reading from it. The two coincide only when the row permutation is an involution, which is why the bug was seed-dependent.

### CODE REVIEW

**Comment by victor-onofre** [comment]
It fixes the error. But while testing this, I have seen another bug.  The solution's output is not sorted by ascending energy. 

`solution[0]` or` bitstrings[0] ` is not the best solution. Something like this happens:

`costs in returned order: [-9.0, -8.0, -8.0, -8.0, -11.0]`
`solution[0].cost = -9.0    best = -11.0`

And counts should be how many times each solution was visited during the run, but it's always 1 for every bitstring. Here is the code to reproduce it:

```
import torch

from qubosolver import Instance, bitstring, matrix, torch_rng
from qubosolver.solvers.classical.simulated_annealing import simulated_annealing

# 3 variables -> only 8 possible bitstrings in the whole search space.
solution = simulated_annealing(
    Instance(matrix.tensor([[-2.0, 1.0, 0.0], [1.0, -3.0, 2.0], [0.0, 2.0, -1.0]])),
    bitstring.zeros(3),
    top_k=4,
    max_iter=5000,  # 5000 steps over 8 states -> revisits are unavoidable
    initial_temp=4.0,
    final_temp=0.05,
    rng=torch_rng(0),
)

print("bitstring   counts   probability")
for single, c in zip(solution, solution.counts.tolist()):
    print(f"  {single.string}         {c}      {single.probability:.4f}")

print()
print("all counts == 1 :", bool(torch.all(solution.counts == 1)))
print("probs uniform   :", bool(torch.allclose(solution.probabilities,
                                               solution.probabilities[0])))
print()
print("5000 annealing steps over 8 possible bitstrings, yet every count is 1.")
print("Real visit frequencies would be large and uneven.")
```

**Review by victor-onofre — APPROVED** [review]
